bin/propolis-server/src/lib/vm.rs

```diff
@@ -578,6 +571,60 @@ impl VmController {
         Ok(controller)
     }
 
+    pub fn stop(self: Arc<VmController>) {
+        let worker_thread = self.worker_thread.lock().unwrap().take();
+
+        let failed = match worker_thread.map(|t| t.join()) {
+            // Worker thread cleanly exited
+            Some(Ok(_)) => false,
+            Some(Err(err)) => {
+                error!(self.log, "VmController worker thread failed"; "err" => ?err);
+                true
+            }
+            None => {
+                error!(self.log, "Active VmController without a worker thread");
+                true
+            }
+        };
+
+        info!(self.log, "Dropping VmController";
+            "strong_refs" => Arc::strong_count(&self),
+            "weak_refs" => Arc::weak_count(&self),
+            "instance_refs" => Arc::strong_count(self.instance()),
+        );
+
+        let api_state = match Arc::try_unwrap(self) {
+            Ok(vm) => {
+                // We should have the only reference to the instance at this point
+                debug_assert_eq!(Arc::strong_count(vm.instance()), 1);
+
+                // Grab the sender side of the watcher for the instance state
+                vm.vm_objects.api_state.clone()
+
+                // Once the VmController (vm) is dropped here, so too will the last
+                // reference to the instance thus cleaning up its VMM handle.
+            }
+            Err(vm) => {
+                error!(vm.log, "VmController has outstanding references");
+                vm.vm_objects.api_state.clone()
+            }
+        };
+
+        // Send a final state monitor message marking the instance as failed
+        // or destroyed. Normally, the existence of this structure implies the
+        // instance of at least one receiver, but at this point everything is
+        // being dropped, so this call to `send` is not safe to unwrap.
+        let gen = api_state.borrow().gen + 1;
+        let _ = api_state.send(ApiMonitoredState {
+            gen,
+            state: if failed {
+                ApiInstanceState::Failed
+            } else {
+                ApiInstanceState::Destroyed
```

Comment:
I could just be missing something, but I'm not sure this totally closes the race--we're publishing `Destroyed` here, but IIUC `self` is still in scope, so the instance still hasn't quite been dropped yet. (We'll also publish `Destroyed` even if we logged an error on line 608. Today I don't think this is much of a problem, since no one in the server module currently uses `VmController` references outside of the server's state lock, but I wonder if we should allow that--it feels a little weird to have this routine's correctness depend on the caller's synchronization scheme.)

Would it make sense to explicitly `destroy` the instance in this path instead (or turn `VmController::instance` into an `Mutex<Option<Arc<Instance>>>` to force everyone to reckon with the possibility that it's been destroyed)?

Reply:
`self` is out of scope at this point as `Arc::try_unwrap` takes it by value. (Though the `Err` case does mean there are other refs out there)

I went back and forth on just calling `Instance::destroy` manually but decided not to at least for now to not paper over leftover references. Sticking the instance behind an Option also meant a cascade of other changes which I suppose isn't the worst.

I'll go ponder on some approaches about how to lay things out so everything doesn't get too tangled.